pyra/hardware.py

```diff
@@ -0,0 +1,422 @@
+
+    This will append a new value to the ``dash_stats['cpu_system']`` list.
+
+    Returns
+    -------
+    float
+        The current system cpu percentage utilized.
+
+    Examples
+    --------
+    >>> update_cpu()
+    """
+    cpu_percent = psutil.cpu_percent(interval=None, percpu=False)
+
+    if initialized:
+        dash_stats['cpu_system'].append(cpu_percent)
+
+    return cpu_percent
+
+
+def update_gpu():
+    """
+    Update dashboard stats for system GPU usage.
+
+    This will create new keys for the ``dash_stats`` dictionary if required, and then append a new value to the
+    appropriate list.
+
+    Nvidia GPUs are fully supported. AMD GPUs are currently only supported on Linux.
+
+    Examples
+    --------
+    >>> update_gpu()
+    """
+    global nvidia_gpus
+    nvidia_gpus = GPUtil.getGPUs()  # need to get the GPUs again otherwise the load does not update
+
+    for gpu in nvidia_gpus:
+        name = f'{gpu.name}-{gpu.id}'
+
+        if initialized:
+            try:
+                dash_stats[f'gpu_{name}']
+            except KeyError:
+                dash_stats[f'gpu_{name}'] = []
+            finally:
+                gpu_load = 100 if gpu.load > 1 else gpu.load * 100  # convert decimal to percentage, max of 100
+                dash_stats[f'gpu_{name}'].append(gpu_load)
+
+    # this is untested
+    amd_counter = 0
+    while amd_counter < amd_gpus:
+        gpu = pyamdgpuinfo.get_gpu(amd_counter)
+
+        name = f'{gpu.name}-{gpu.gpu_id}'
+
+        if initialized:
+            try:
+                dash_stats[f'gpu_{name}']
+            except KeyError:
+                dash_stats[f'gpu_{name}'] = []
+            finally:
+                gpu_load = 100 if gpu.query_load() > 100 else gpu.query_load()  # max of 100
+                dash_stats[f'gpu_{name}'].append(gpu_load)
+
+        amd_counter += 1
+
+
+def update_memory():
+    """
+    Update dashboard stats for system memory usage.
+
+    This will append a new value to the ``dash_stats['memory_system']`` list.
+
+    Returns
+    -------
+    float
+        The current system memory percentage utilized.
+
+    Examples
+    --------
+    >>> update_memory()
+    """
+    memory_percent = psutil.virtual_memory().percent
+
+    if initialized:
+        dash_stats['memory_system'].append(memory_percent)
+
+    return memory_percent
+
+
+def update_network():
+    """
+    Update dashboard stats for system network usage.
+
+    This will append a new values to the ``dash_stats['network_received']`` and ``dash_stats['network_sent']``
+    lists.
+
+    Returns
+    -------
+    tuple
+        A tuple of the received and sent values as a difference since the last update.
+
+    Examples
+    --------
+    >>> update_network()
+    """
+    global initialized
+    global network_recv_last
+    global network_sent_last
+
+    net_recv_current = psutil.net_io_counters().bytes_recv / 1000000  # convert bytes to mb
```

Comment:
You could have one call to psutil.net_io_counters(), store the result, and then just access the 2 attributes you are interested in. That way, you don't have to generate 2 objects to access information that both exist on either object.

Reply:
Good call!